src/main/java/uk/ac/ic/wlgitbridge/bridge/context/ContextStore.java

```diff
@@ -0,0 +1,132 @@
+package uk.ac.ic.wlgitbridge.bridge.context;
+
+import uk.ac.ic.wlgitbridge.util.Log;
+import java.util.HashMap;
+
+/**
+ * A singleton that manages Project Contexts in the application.
+ * Contexts are wrappers around a ProjectLock, and any other
+ * project-specific resources that need to be managed during
+ * an "operation" on a project.
+ *
+ * Contexts are kept in a ThreadLocal variable, so they can
+ * be re-used once they have been created.
+ *
+ * The intended use is to open a context at the 'top' of a
+ * git operation (push, pull, swap), and lock the project.
+ * Then, deeper in the call-stack the same context can be
+ * retrieved and used.
+ *
+ * (In practice, this enables us to use a Postgres connection
+ * to lock a row in the database, and hold a transaction open,
+ * to be used further down the call stack.)
+ */
+public class ContextStore {
+  private ProjectContextFactory projectContextFactory;
+  static private ContextStore instance;
+  private boolean shuttingDown;
+
+  private ThreadLocal<HashMap<String, ProjectContext>> localContexts;
```

Comment:
Does this need to be a `HashMap`?

My understanding is that the thread-local variable should just be a context, as the thread should only be working on one project at a time. The thread-local mechanism is to get hold of the context for the current request, without passing it all the way down through the stack. I can't see why we would need to have multiple ones for different projects existing at the same time. WDYT?

Reply:
Discussed in call:

Actually, this is correct. We sometimes need to lock multiple projects at once, when we are migrating things around.

Reply:
I'm still confused about how the thread-local locks prevent multiple threads from obtaining a lock on the same project. I trust that it's been thoroughly discussed, but I wouldn't mind an explanation when you have some time :)